Sites: Receive site updates when sites-list fetch completes

aduth opened the pull request:
Related: #7775 
Fixes #10450

This pull request seeks to enable Redux to receive site updates which occur in sites-list as a result of polling site fetches. In hopes of minimizing syncing issues between `sites-list` and Redux, it also modifies the site controller such that a site is only received into Redux state if not already tracked, under the assumption that otherwise the data may not necessarily be fresh if changes had been made in Redux state.

__Implementation notes:__

My first attempt was to hook into the sites `change` event handler, but this occurs more than just sites being synced from REST API, defeating the purpose by causing the more recent Redux state to still be overridden. The worry that remains is if we have changes that are considered newer in sites-list than in Redux but won't be received because of the changes to the sites controller.

Refactoring of reducer sites received action handlers includes consolidation of logic and improved handling of unchanged state.

__Testing instructions:__

Repeat testing instructions from #10450, notably assigning or removing a site icon and leaving and reentering the sites settings page, observing that the intended action sticks as expected.

dmsnell commented on client/my-sites/controller.js:
```diff
@@ -169,7 +170,9 @@ module.exports = {
 			siteStatsStickyTabActions.saveFilterAndSlug( false, selectedSite.slug );
-			context.store.dispatch( receiveSite( selectedSite ) );
+			if ( ! getSite( context.store.getState(), selectedSite.ID ) ) {
+				context.store.dispatch( receiveSite( selectedSite ) );
```
This is an odd conditional. Would we be getting a `selectedSite` if it didn't already exist in the state tree?

aduth replied:
>This is an odd conditional. Would we be getting a `selectedSite` if it didn't already exist in the state tree?

I added some clarification to my comment at https://github.com/Automattic/wp-calypso/pull/10477#issuecomment-271590520 . Essentially we want to receive the site object only if we don't already have it tracked in Redux state. Otherwise we risk receiving in stale data if changes have since been made to the Redux state site.

dmsnell commented on client/state/current-user/reducer.js:
```diff
@@ -60,28 +61,35 @@ export const currencyCode = createReducer( null, {
-		}
+export const capabilities = withSchemaValidation(
+	capabilitiesSchema,
+	( state = {}, action ) => {
```
I thought I had already left a comment here but GitHub isn't showing it to me, so please pardon me if I'm redundant here (it was from my phone, so maybe that's why)

it's rare that I think it's a good idea to define a function inside of the argument list of another function call. in fact, I usually consider it a bad idea (such as inlining the mapping functions inside of `connect()`).

if we pull this out to its own named function (and export) then it's easier to test in isolation and I think it's also easier to see what's going on, that we have a raw reducer and then we have a composed version of it which adds schema validation.

last time I saw the discussion, you and @gwwar were both in favor of inlining inside of `connect()`, although I think we can agree that we're all using a subjective line there or else we would also define our React components inside of the call to `connect()`.

in this case, I feel the line has been crossed again just because we can't test the reducer in isolation or recompose it in other ways.

hope this wasn't too rantful :smile:

aduth replied:
>it's rare that I think it's a good idea to define a function inside of the argument list of another function call.

I think the difference in how I view this is in reading the code, I'd like to consider this to be one individual unit: a reducer (which happens to be enhanced through utilities). Especially in a reducer file, if we were to split this up, I've suddenly lost confidence in being able to identify / differentiate reducers in the file (yes, they may technically still all be reducers).

While you're correct that the function argument couldn't be tested in isolation in its current form, the fact of the matter is that I have no intention of testing it anyways, so pragmatically I'll accept the caveat and happily accept willingness to make minor adjustments to the code if I ever change my mind in the future.

dmsnell commented on client/state/current-user/reducer.js:
```diff
@@ -60,28 +61,35 @@ export const currencyCode = createReducer( null, {
+		}
 
-		return Object.assign( {}, state, siteCapabilities );
+		return state;
```
Is this a good place for set-membership check of action type and early abort?

```js
if ( ! includes( [ SITE_RECEIVE, SITES_RECEIVE, SITES_UPDATE ], action.type ) {
	return state;
}

const sites = …
```

aduth replied:
>Is this a good place for set-membership check of action type and early abort?

Yeah, aside from breaking convention of `switch` and perhaps requiring some revisions in the future if we start handling additional action types, I think this could help avoid a couple indents.

dmsnell commented on client/state/sites/enhancer.js:
```diff
@@ -0,0 +1,31 @@
+	const sites = require( 'lib/sites-list' )();
+	sites.sync = flow( sites.sync.bind( sites ), () => {
+		store.dispatch( receiveSiteUpdates( sites.get() ) );
+	} );
```
on style, can we leave on partial function per line? I find it much more readable that way…

```js
sites.sync = flow(
	sites.sync.bind( sites ),
	() => store.dispatch( receiveSiteUpdates( sites.get() ) ),
);
```

also (and you may disagree) but I find the `{}`s distracting and unnecessary here

aduth replied:
>on style, can we leave on partial function per line? I find it much more readable that way…

Sure.

>also (and you may disagree) but I find the {}s distracting and unnecessary here

For whatever reason, I'm happy without the curly brackets in the inline case, but multi-line triggers an internal desire to see the start/end more clearly.

aduth replied:
>on style, can we leave on partial function per line? I find it much more readable that way…
>also (and you may disagree) but I find the {}s distracting and unnecessary here
  
Both updated in rebased 9cc2607.

dmsnell commented on client/state/sites/reducer.js:
```diff
@@ -98,18 +99,38 @@ export function items( state = {}, action ) {
+			return reduce( sites, ( memo, site ) => {
+				// If we're not already tracking the site upon an update, don't
+				// merge into state (TODO: Consider dropping condition once
+				// sites-list abolished)
```
if we put the TODO on it's own line it will end up rendering better in tools which highlight and track TODOs

aduth replied:
>if we put the TODO on it's own line it will end up rendering better in tools which highlight and track TODOs

Can do.

aduth replied:
>if we put the TODO on it's own line it will end up rendering better in tools which highlight and track TODOs
  
Updated in rebased 9cc2607.

dmsnell commented on client/state/sites/reducer.js:
```diff
@@ -98,18 +99,38 @@ export function items( state = {}, action ) {
-				memo[ site.ID ] = pick( site, VALID_SITE_KEYS );
+		case SITES_UPDATE:
+			// Normalize incoming site(s) to array
+			const sites = action.site ? [ action.site ] : action.sites;
```
let's be careful here. the case statement doesn't create a new scoping block and so `const sites` can only exist in this branch.

aduth replied:
>let's be careful here. the case statement doesn't create a new scoping block and so `const sites` can only exist in this branch.

Yeah, unfortunate really, though ESLint will capture this (as an error, if I'm not mistaken). In other cases I've used plain [JavaScript blocks](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/block) to create a scope for the case, like:

```js
case SITES_UPDATE: {
	const sites = action.site ? [ action.site ] : action.sites;
	// ...
}
```

Though given that this is not a common pattern, may be more confusing than it helps.

dmsnell commented on client/state/sites/reducer.js:
```diff
@@ -98,18 +99,38 @@ export function items( state = {}, action ) {
+				memo[ site.ID ] = transformedSite;
 				return memo;
-			}, {} );
+			}, initialMemo );
```
This is looking good, but I find the term `memo` confusing. it looks like's it's being used to represent the previous sites state. did the term come from "memoization" or "memory"?

aduth replied:
>This is looking good, but I find the term `memo` confusing. it looks like's it's being used to represent the previous sites state. did the term come from "memoization" or "memory"?

I've used for some time, but if I were to guess, I might have picked it up from Underscore.js's documentation back in the day: http://underscorejs.org/#reduce

Doesn't seem like [too uncommon a term](https://www.google.com/search?q=array+reduce+memo) and you're probably right it's a convenient derivation of one of either memory or memoization.

What would your suggested alternative be?

gwwar replied:
`accumulator` or `initialValue` is pretty common. In this case though, what about `lastState`?

aduth replied:
>`accumulator` or `initialValue` is pretty common. In this case though, what about `lastState`?
  
Oof, I thought I'd read this as `initialState` when I'd updated in rebased 9cc2607. Though `lastState` is not entirely accurate, since it's not so much what the previous state is, but the basis for what the next state will be. Maybe something like `initialNextState`? Quite a mouthful.

gwwar commented:
If I understand the issue correctly:

1. We modify site state that takes some time to update on the server
2. SiteList/QuerySite can wipe this data either on an interval, or render with stale data
3. To avoid this we compare last given action state (deep equals on each site obj) to avoid updating site state if nothing changed in the update.

I'm getting some odd behavior following steps in 10450. Waiting for the sites list poll, makes the icon popin (including just staying on /settings). Clicking around will clear out the site icon, but if you wait for the next poll, it will still pop in with the old icon. 
![popin](https://cloud.githubusercontent.com/assets/1270189/21790201/729d61cc-d68f-11e6-94cc-5514c2f45ecf.gif)

Let me know if you can't repo.

A few questions:
- Do we need any changes to the`requestSite()` and `requestSites()` behavior that's used in QuerySites component? 
- What happens when unrelated data changes in the site object, but our changes haven't gone through yet?

aduth commented:
Thanks for the reviews @dmsnell and @gwwar !

@gwwar :

>To avoid this we compare last given action state (deep equals on each site obj) to avoid updating site state if nothing changed in the update.

This is not so much part of the problem I sought to fix, but rather an optimization discovered through the changes. Your first two points are right though, we need to address the issue where `sites-list` variation of the site is out of date with changes that have since occurred in Redux state. Previously we'd received the site into Redux state from `sites-list` on the site route controller which could override those newer changes. We still need this behavior since this is the only way that sites data gets into Redux state in the first place, but once we have the site in Redux state, we don't necessarily need to continue receiving it in. However, if we'd like for the sites state to stay up-to-date, we can hook into the `sites-list` fetching mechanism and receive into Redux state directly at that point. This is not too much unlike #7775, except we only track site updates received from the polling if we'd already had the site in state previously.

>Clicking around will clear out the site icon, but if you wait for the next poll, it will still pop in with the old icon.

This sounds like it shouldn't be happening, especially when you're staying on the same page. It's almost as if the data received from the server is still stale, even though it's freshly received? Maybe some bad caching with the icon server-side. I'll take a look.

>Do we need any changes to the `requestSite()` and `requestSites()` behavior that's used in QuerySites component?

Are we ever requesting sites from query components? In any case, I don't think should be concern for the changes here, though is problematic for pages still relying on `sites-list`, since the data would be stale.

>What happens when unrelated data changes in the site object, but our changes haven't gone through yet?

Can you clarify which site object you mean to be changed here?

gwwar commented:
Thanks for the clarification @aduth 

> Are we ever requesting sites from query components?

Yes, some components do use `<QuerySites siteId={ siteID } />'. We even have a single usage of `<QuerySites />` that snuck in a while back to handle a corner case. 

> Can you clarify which site object you mean to be changed here?

I was just thinking of a hypothetical scenario on how a pop in effect can occur. Maybe something like:

1. Optimistically update site state, while triggering an update on the server
2. Polled/Queried data comes back before changes go through, wiping out the local optimistic updates
3. On next poll/query, state updates to what we expect

aduth commented:
>I was just thinking of a hypothetical scenario on how a pop in effect can occur. Maybe something like:

Ah yes, this could certainly be an issue, and one I don't really have a good answer to. Even in a scenario where we're fully migrated to Redux state, if we were to keep polling around, this seems like it would continue to be a problem. Maybe at that point we don't have polling and assume that our local state is up-to-date? But in that case we'd not account for changes made by other users / in separate browser sessions.

aduth commented:
Looping in @Copons , since the changes here also enable DevTools in production when detected in the browser (discussed recently). Not quite intentional, but I personally think it'd be a fine change.

aduth commented:
One of my remaining worries with this branch was that there's still legacy behaviors which apply changes to individual sites in the `sites-list`, and those would have no longer have been reflected until the next complete sites fetch.

In b4977cd, I sought to account for these changes by overriding the prototype of `site.set` (_I know, hacks upon hacks_). This way, if a legacy flow calls upon `site.set` with a set of attributes, those attributes should also be reapplied on top of what we know to be the Redux state site object. With refactoring in 1b68f3b869c049c6d86bc8d2a053440eee63b6f9, we should expect that the reducer determines whether there are in-fact any changes (deeply) before modifying state.

You can test the behavior by observing that the Jetpack [received plugin updates](https://github.com/Automattic/wp-calypso/blob/006bdef6b83351cd566ad55ea5114379da4219ae/client/lib/site/jetpack.js#L62-L75) triggering a `set` are mirrored to sites state via `SITE_RECEIVE` if the site is tracked in state (selected sites). Other examples I've found of legacy behavior include a `plans` attribute being updated for a site in `sites-list` after having made a purchase.

I've also rebased the branch and added some clarifying comments to the sites controller logic in 0c6b09749083190c3a989fcd53945363b1c4d245.

@dmsnell @gwwar How are you feeling about these changes? I personally feel that something like this is necessary as we rely more on Redux sites state, because currently we blindly receive the `sites-list` last known site on every site-specific screen, which will become more obviously out-of-date as we continue to port site behaviors to Redux.

gwwar commented:
> How are you feeling about these changes? I personally feel that something like this is necessary as we rely more on Redux sites state, because currently we blindly receive the sites-list last known site on every site-specific screen, which will become more obviously out-of-date as we continue to port site behaviors to Redux.

I'm not opposed to this. In the future I was hoping to make reverse the information flow, where the redux tree is the source of truth and dispatch updates back to SitesList while we remove SitesList references. I haven't had enough bandwidth recently to tackle this though.

Also, I'm running into the same behavior I described in https://github.com/Automattic/wp-calypso/pull/10477#issuecomment-271460873 If this PR fixes other cases, could you add steps to test?

aduth commented:
>In the future I was hoping to make reverse the information flow, where the redux tree is the source of truth and dispatches updates back to SitesList while we remove SitesList references.

Yeah, this seems appealing, but I wonder if it discourages us from updating those components which continue to rely on the legacy behaviors. I think both approaches could eventually work out in the end. The changes here do at least give us a path toward incrementally picking off and migrating the few pieces that still make changes to sites-list sites directly. In fact, I'm curious if a few of those I've found—particularly plans and plugins—wouldn't be much trouble to "finish".

>Also, I'm running into the same behavior I described in #10477 (comment) If this PR fixes other cases, could you add steps to test?

I'll test this now, still unsure why it'd be happening.

Will also rebase again; looks like a minor import conflict on the sites controller.

aduth commented:
@gwwar It appears you stumbled into quite an obscure bug in the sites-list implementation that's likely existed since the beginning of time.

In particular, with the flow you described, the site returned from the REST API changes from including an `icon` property to not including the property (which is meant to indicate that the site does not have an icon).

However, after `sites-list` is first initialized, subsequent changes for fetched sites are applied as a patch to the existing site object via `Site.prototype.set`. Therefore, the lack of `icon` is not reflected in the new site object because it's assumed to not have been modified:

- https://github.com/Automattic/wp-calypso/blob/8c150e1/client/lib/sites-list/list.js#L204
- https://github.com/Automattic/wp-calypso/blob/8c150e1/client/lib/site/index.js#L63-L81

In my mind, if we receive updates for sites, it would seem to make sense to replace outright the version we're currently tracking, rather than apply as a patch update. It could be that the original implementation was written as such to try to more easily detect when a change in-fact occurs (the return value of `set`) to avoid needlessly emitting `change` events. The `icon` property is a bit of an outlier in how it can be omitted from the response, so it's not unreasonable to assume that its behavior could have been overlooked.

I might suggest that we consider revising `sites-list` as part of a separate pull request, since the bug exists independent of the changes here.

- 748-gh-calypso-pre-oss
- 1182-gh-calypso-pre-oss

aduth commented:
Rebased again to resolve latest merge conflicts.

#10713 includes a fix for the [aforementioned behavior](https://github.com/Automattic/wp-calypso/pull/10477#issuecomment-273295260). I will plan to merge both that and this branch in the morning assuming there are no outstanding concerns.
